K2Bridge/Visitors/HistogramVisitor.cs

```diff
@@ -0,0 +1,40 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+// See LICENSE file in the project root for full license information.
+
+namespace K2Bridge.Visitors
+{
+    using System;
+    using K2Bridge.Models.Request.Aggregations;
+
+    /// <content>
+    /// A visitor for the <see cref="HistogramAggregation"/> element.
+    /// </content>
+    internal partial class ElasticSearchDSLVisitor : IVisitor
+    {
+        /// <inheritdoc/>
+        public void Visit(HistogramAggregation histogramAggregation)
+        {
+            Ensure.IsNotNull(histogramAggregation, nameof(histogramAggregation));
+            EnsureClause.StringIsNotNullOrEmpty(histogramAggregation.Metric, nameof(histogramAggregation.Metric));
+            EnsureClause.StringIsNotNullOrEmpty(histogramAggregation.Field, nameof(histogramAggregation.Field));
+
+            histogramAggregation.KustoQL = $"_data | {KustoQLOperators.Summarize} " + histogramAggregation.SubAggregationsKustoQL +
+            $"{histogramAggregation.Metric} by {EncodeKustoField(histogramAggregation.Key + "%" + histogramAggregation.Keyed)} = ";
+
+            var interval = Convert.ToInt32(histogramAggregation.Interval);
+            var field = EncodeKustoField(histogramAggregation.Field, true);
+
+            histogramAggregation.KustoQL += $"bin({field}, {interval})";
+
+            // todatetime is redundent but we'll keep it for now
+            histogramAggregation.KustoQL += $"{KustoQLOperators.CommandSeparator}{KustoQLOperators.OrderBy} {EncodeKustoField(histogramAggregation.Key + "%" + histogramAggregation.Keyed)} asc";
+            if (histogramAggregation.HardBounds != null)
+            {
+                var min = Convert.ToInt32(histogramAggregation.HardBounds.Min) - interval;
+                var max = Convert.ToInt32(histogramAggregation.HardBounds.Max) + interval;
+                histogramAggregation.KustoQL += $" | where {EncodeKustoField(histogramAggregation.Key + "%" + histogramAggregation.Keyed)} between (" + min + " .. " + max + ") ";
```

Comment:
1. Use the KustoQLOperators class instead of 'where' directly.
2. It looks like you aggregate and then filter. Can't you filter before aggregating?

Reply:
1. Done
2. The query looks like :
```kusto
let _data = database("dev").kibana_data_flights 
| where (['timestamp'] >= todatetime("2007-01-03T14:15:49.6730000Z") and ['timestamp'] <= todatetime("2022-01-03T14:15:49.6730000Z"));
(_data | summarize count() by ['2'] = bin(['AvgTicketPrice'], 10)
| order by ['2'] asc | where ['count_'] between (50 .. 150) | as aggs);
(_data | count | as hitsTotal);
(_data | limit 0 | as hits)
```
The where filtering came after the column definition. If we want to filter before aggregating, how the query looks like ?

Reply:
That's the confusion part @nebrass.
Not sure this example query is what gets generated by the code:
L34: histogramAggregation.KustoQL += $"{KustoQLOperators.CommandSeparator}{KustoQLOperators.OrderBy} {**histogramKey**} asc";
L39: histogramAggregation.KustoQL += $" | {KustoQLOperators.Where} {**histogramKey**} between ({min} .. {max}) ";
Notice that you use the same _histogramKey_ field in the code on both lines but in your example it is ['2'] and ['count_']...

At any case - you should filter and then order (if you can).

Reply:
It's the same query, just the column name evolved to hold a keyed flag for the aggregation:
```kusto
let _data = database("dev").kibana_data_flights 
| where (['timestamp'] >= todatetime("2007-01-03T14:15:49.6730000Z") and ['timestamp'] <= todatetime("2022-01-03T14:15:49.6730000Z"));
(_data | summarize count() by ['2%False'] = bin(['AvgTicketPrice'], 10)
| order by ['2%False'] asc | where ['count_'] between (50 .. 150) | as aggs);
(_data | count | as hitsTotal);
(_data | limit 0 | as hits)
```
This is exactly the build query

Reply:
You have the wrong variable in "where ['count_'] between (50 .. 150) | as aggs);".
Instead of {histogramKey} in
`                histogramAggregation.KustoQL += $" | {KustoQLOperators.Where} {histogramKey} between ({min} .. {max}) ";`
it should be {BucketColumnNames.Count}

The reason it wasn't caught was because you don't have any tests with hard_bounds (nor extended_bounds), so it was never caught.


Also Tamir said to move that if before the order by like so:
```c#
  if (histogramAggregation.HardBounds != null)
  {
                var min = Convert.ToInt32(histogramAggregation.HardBounds.Min) - interval;
                var max = Convert.ToInt32(histogramAggregation.HardBounds.Max) + interval;
                histogramAggregation.KustoQL += $" | {KustoQLOperators.Where} {BucketColumnNames.Count} between ({min} .. {max}) ";  
}
 histogramAggregation.KustoQL += $"{KustoQLOperators.CommandSeparator}{KustoQLOperators.OrderBy} {histogramKey} asc";
```

So the final query will filter before ordering:
```kusto
let _data = database("dev").kibana_data_flights 
| where (['timestamp'] >= todatetime("2007-01-03T14:15:49.6730000Z") and ['timestamp'] <= todatetime("2022-01-03T14:15:49.6730000Z"));
(_data | summarize count() by ['2%False'] = bin(['AvgTicketPrice'], 10)
| where ['count_'] between (50 .. 150) | order by ['2%False'] asc |  as aggs);
(_data | count | as hitsTotal);
(_data | limit 0 | as hits)
```

Reply:
After checking the demo elastic, it's the bin() value hold in histogramKey that needs to be delimited by the hardbounds:
<img width="3007" alt="Capture d’écran 2022-01-10 à 11 05 26" src="https://user-images.githubusercontent.com/2045375/148748658-d28c00e5-f5fe-4a7b-9986-24b8b51c567c.png">